### PR TITLE
Build on windows using appveyor

### PR DESCRIPTION
**Pull request by pierreinglebert**

### CODE REVIEW

**Comment by XhmikosR** [comment]
Can't you just test with the default VS and stuff and simplify this? It's too bloated... Also, keep in mind AppVeyor is always slow for free plans, so you know...

**Comment by pierreinglebert** [comment]
It's for node-pre-gyp, many users on windows doesn't have a proper VS installation to build native modules.
For a while I built them by hand and published it on a s3 bucket, switching to a x86 & x64 VM and with 0.10 & 0.12 node, if I can automate this...

**Comment by XhmikosR** [comment]
Sure, we can definitely do this. But I still believe you should use VS2013 which is the default VS AppVeyor is using. Also, make use of `"%VS120COMNTOOLS%..\..\VC\vcvarsall.bat"` instead of specifying the whole path; it's always in PATH.

Just try to have a working version of this merged and we can simplify and improve this :)

**Comment by XhmikosR** [comment]
@pierreinglebert: I'm working on switching to shelljs to have a cross platform solution. Apparently, `npm i` fails for me on Windows 7 64-bit.  https://gist.githubusercontent.com/XhmikosR/23ec9f8993c0d1aff769/raw/zopli.log

I remember this worked yesterday so it must be something new.

**Comment by pierreinglebert** [comment]
Really strange as it yells about the redifinition of 'EscapableHandleScope::~ EscapableHandleScope' which is in v8 from node 0.11.

The only change made recently in c++ source is :
https://github.com/pierreinglebert/node-zopfli/commit/2ddfa44a858ca451631b3dfc8a0e66e0159db71d#diff-f8cbcedd1ac9210cbf41ed193f65774b

**Comment by XhmikosR** [comment]
I'm on node.js 0.12.2; here you test only 0.10. We really need this PR merged for testing first, and deployment after it works.

**Comment by XhmikosR** [comment]
Note, that currently AppVeyor is using VS2010 it seems.

**Comment by pierreinglebert** [comment]
nope, look at the last build, appveyor builds are made on 0.10,0.12,iojs on both x86/x64

**Comment by XhmikosR** [comment]
Yeah, I edited my message :P

**Comment by XhmikosR** [comment]
I don't have any other VS version to test, so the only difference is that. Can you try VS2013 and not VS2010?

**Comment by XhmikosR** [comment]
Some hints for Appveyor:

```yml
matrix:
  fast_finish: true

cache:
  - '%APPDATA%\npm\node_modules -> package.json'    # global npm modules
  - '%APPDATA%\npm-cache -> package.json'           # npm cache
  - 'node_modules -> package.json'                  # local npm modules
```

Not sure if it's worth caching `%LOCALAPPDATA%\.node-gyp` too.

**Comment by pierreinglebert** [comment]
There is no oses with VS2013, only 2014 and 2015, Azure and MinGW

**Comment by XhmikosR** [comment]
I'm pretty sure there is because I use it  https://ci.appveyor.com/project/SubtitleEdit/subtitleedit/branch/master.

**Comment by pierreinglebert** [comment]
From the appveyor settings panel for oses :
Default Azure
MinGW
Previous MinGW
Previous Unstable
Previous Visual Studio 2014 CTP4
Previous Visual Studio 2015 CTP
Previous Visual Studio 2015 CTP 6
Previous Visual Studio 2015 Preview
Previous Windows Server 2012 R2
SharePoint 2013
Unstable
Visual Studio 2014 CTP4
Visual Studio 2015 CTP
Visual Studio 2015 CTP 6
Visual Studio 2015 Preview
Windows Server 2012 R2

**Comment by XhmikosR** [comment]
Don't set an OS and just call the batch file I posted a few lines above.

**Comment by XhmikosR** [comment]
@pierreinglebert: use this scheme and remove `msvs_toolset`; that way you will use VS2013 which is the default anyway.

```bat
"%VS120COMNTOOLS%..\..\VC\vcvarsall.bat"
```

File: package.json [hunk 1]
```diff
@@ -50,7 +50,7 @@
   ],
   "scripts": {
     "install": "node-pre-gyp install --fallback-to-build --build-from-source",
-    "test": "make test"
+    "test": "mocha test"
```
Comment (XhmikosR) [inline package.json]: AppVeyor should have `make` and stuff via msysGit in PATH already. Did you try it?

**Comment by pierreinglebert** [comment]
seems it needed  `--msvs_version=2013` option on npm install

make wasnt available, that's why I set mocha to check that tests passed.

**Comment by XhmikosR** [comment]
Weird about make. Anyway, after my shell-js patch we won't need to worry about that.

File: appveyor.yml [hunk 2]
```diff
@@ -0,0 +1,71 @@
+environment:
+  node_pre_gyp_accessKeyId:
+    secure: sNiYns5iJ4x7UzU3Vwt6cwjmSKlJnVYTNPj/wJxtHes=
+  node_pre_gyp_secretAccessKey:
+    secure: aqsNf2cyJ0hpprZZvwJ3bLeUeFjZ0OhbJSulfs9VLA4rqmADnnbU//aL/NBoXQQJ
+  matrix:
+    - nodejs_version: "0.10"
+    - nodejs_version: "0.12"
+    - nodejs_version: "1.0"
```
Comment (XhmikosR) [inline appveyor.yml]: You should use `1` otherwise `1.0` gets pretty old very fast :P

File: appveyor.yml [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+  #     msvs_toolset: 12
+  #   - nodejs_version: "1.8"
+  #     platform: x64
+  #     msvs_toolset: 12
+
+matrix:
+  fast_finish: true
+
+cache:
+  - '%APPDATA%\npm\node_modules -> package.json'    # global npm modules
```
Comment (XhmikosR) [inline appveyor.yml]: You can remove this line; you don't seem to install any modules globally.

File: appveyor.yml (hunk 2 above)
Comment (pierreinglebert) [inline appveyor.yml]: It's already outdated, 2.0 is coming ^^

Comment (XhmikosR) [inline appveyor.yml]: Better than 1.0 :)

File: appveyor.yml [hunk 4]
```diff
@@ -0,0 +1,52 @@
+environment:
+  node_pre_gyp_accessKeyId:
+    secure: sNiYns5iJ4x7UzU3Vwt6cwjmSKlJnVYTNPj/wJxtHes=
+  node_pre_gyp_secretAccessKey:
+    secure: aqsNf2cyJ0hpprZZvwJ3bLeUeFjZ0OhbJSulfs9VLA4rqmADnnbU//aL/NBoXQQJ
+  matrix:
+    - nodejs_version: "0.10"
+    - nodejs_version: "0.12"
+    - nodejs_version: "1.8"
```
Comment (XhmikosR) [inline appveyor.yml]: That is the gain of `1`; it will install the latest.

File: appveyor.yml (hunk 2 above)
Comment (pierreinglebert) [inline appveyor.yml]: I set 1.8 (which will be the last 1.x), the 1.0 came from a copy paste from appveyor documentation ^^

File: appveyor.yml [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+  - npm -v
+  # Build
+  - npm install --build-from-source --msvs_version=2013
+  - npm test || true
+  - node-pre-gyp package %BUILD_ARGS%
+  # make commit message env var shorter
+  - SET CM=%APPVEYOR_REPO_COMMIT_MESSAGE%
+  - ps: if($env:APPVEYOR_REPO_COMMIT_MESSAGE.ToLower().Contains('[publish binary]')) { node-pre-gyp --msvs_version=2013 publish }
+
+build: OFF
```
Comment (XhmikosR) [inline appveyor.yml]: nitpicking but maybe use lowercase `off` :)

File: appveyor.yml [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+cache:
+  - '%APPDATA%\npm-cache -> package.json'           # npm cache
+  - 'node_modules -> package.json'                  # local npm modules
+
+install:
+  - ps: Install-Product node $env:nodejs_version
+  # get zopfli submodule
+  - git submodule update --init --recursive
+  # add local node to path (since we install it for msvs_toolset == 14)
+  - SET PATH=%CD%;%PATH%;
```
Comment (XhmikosR) [inline appveyor.yml]: Is this still needed? The comment seems outdated though :P

File: appveyor.yml [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+  # add local node-pre-gyp dir to path
+  - SET PATH=node_modules\.bin;%PATH%
```
Comment (XhmikosR) [inline appveyor.yml]: This might also not be needed since npm searches there by default.

File: appveyor.yml [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+  # use 64 bit python if platform is 64 bit
+  - if "%platform%" == "x64" set PATH=C:\Python27-x64;%PATH%
+  - SET ARCHPATH=
```
Comment (XhmikosR) [inline appveyor.yml]: This and the next line seems redundant now.

**Comment by XhmikosR** [comment]
I hope you don't mind my comments; I really want this to work good on Windows since I'm a mainly Windows user myself :)

File: appveyor.yml [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+  # Print some informations
+  - node -v
+  - node -e "console.log(process.argv,process.execPath)"
+  - npm -v
+  # Build
+  - npm install --build-from-source --msvs_version=2013
+  - npm test || true
+  - node-pre-gyp package %BUILD_ARGS%
+  # make commit message env var shorter
+  - SET CM=%APPVEYOR_REPO_COMMIT_MESSAGE%
```
Comment (XhmikosR) [inline appveyor.yml]: Is this still needed?

File: appveyor.yml [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+  - git submodule update --init --recursive
+  # add local node-pre-gyp dir to path
+  - SET PATH=node_modules\.bin;%PATH%
+  # use 64 bit python if platform is 64 bit
+  - if "%platform%" == "x64" set PATH=C:\Python27-x64;%PATH%
+  - if %platform% == x64 CALL "%VS120COMNTOOLS%..\..\VC\vcvarsall.bat" amd64
+  - if %platform% == x86 CALL "%VS120COMNTOOLS%..\..\VC\vcvarsall.bat" amd64_x86
+  # Print some informations
+  - node -v
+  - node -e "console.log(process.argv,process.execPath)"
```
Comment (XhmikosR) [inline appveyor.yml]: Is this still needed?

File: appveyor.yml [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+  # get zopfli submodule
+  - git submodule update --init --recursive
+  # add local node-pre-gyp dir to path
+  - SET PATH=node_modules\.bin;%PATH%
+  # use 64 bit python if platform is 64 bit
+  - if "%platform%" == "x64" set PATH=C:\Python27-x64;%PATH%
+  - if %platform% == x64 CALL "%VS120COMNTOOLS%..\..\VC\vcvarsall.bat" amd64
+  - if %platform% == x86 CALL "%VS120COMNTOOLS%..\..\VC\vcvarsall.bat" amd64_x86
+  # Print some informations
+  - node -v
```
Comment (XhmikosR) [inline appveyor.yml]: Small typo:"information". Also, how about just doing `node -v && npm -v`?

**Comment by pierreinglebert** [comment]
@XhmikosR any idea why appveyor doesnt like [this](https://ci.appveyor.com/project/pierreinglebert/node-zopfli/build/67/job/yqf4m7pg8g52xw8d) ? 
It worked btw.

**Comment by XhmikosR** [comment]
Maybe the newer node-pre-gyp broke it?

BTW, you can do this for multiline ps scripts

```yml
  - ps: >-
      if ($env:APPVEYOR_REPO_COMMIT_MESSAGE.ToLower().Contains("[publish binary]")) {
        node_modules\.bin\node-pre-gyp --msvs_version=2013 package publish
      }
```

File: appveyor.yml [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+cache:
+  - '%APPDATA%\npm-cache -> package.json'           # npm cache
+  - 'node_modules -> package.json'                  # local npm modules
+
+install:
+  - ps: Install-Product node $env:nodejs_version
+  # get zopfli submodule
+  - git submodule update --init --recursive
+  # add local node-pre-gyp dir to path
+  - SET PATH=node_modules\.bin;%PATH%
```
Comment (XhmikosR) [inline appveyor.yml]: These 2 lines seem redundant since you specify the full path in the PS command.

File: appveyor.yml [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+  - if %platform% == x86 CALL "%VS120COMNTOOLS%..\..\VC\vcvarsall.bat" amd64_x86
+  # Print some information
+  - node -v
+  - npm -v
+  # Build
+  - npm install --build-from-source --msvs_version=2013
+  - npm test
+  - ps: >-
+      if ($env:APPVEYOR_REPO_COMMIT_MESSAGE.ToLower().Contains("[publish binary]")) {
+        cmd /c 'node_modules\.bin\node-pre-gyp --msvs_version=2013 package publish 2>&1'
```
Comment (XhmikosR) [inline appveyor.yml]: Remove `2>&1` otherwise you won't see any errors :P

**Comment by XhmikosR** [comment]
@pierreinglebert: can you try moving the publish command to package.json scripts and use it like `npm run publish` or something like that? Maybe it's an issue due to that like I had with istanbul.

**Comment by pierreinglebert** [comment]
I see errors, they are put in stdout ^^ and when it fails, it exit(1) so appveyor knows something got wrong, I'll try to put it in package.json.

In the meantime, I have another question for you, the x64 env build a x86 binary... Do you know why ?

**Comment by XhmikosR** [comment]
Node used must be 32 bit. Can you add back that snippet to get the node
platform in order to confirm this? You can remove it after that again.

**Comment by XhmikosR** [comment]
Or it's  due to the msvs_version you pass. Maybe it's different for 64 bit
Windows?

**Comment by pierreinglebert** [comment]
It's okay now !

**Comment by XhmikosR** [comment]
So what was the cause?

**Comment by pierreinglebert** [comment]
`  - ps: Install-Product node $env:nodejs_version $env:platform`
There was no platform...
Now, he doesn't care of the allow_failures...

**Pull request merged**